### ISSUE TITLE
Enhancement: [restrict-template-expressions, restrict-plus-operands/] should allow certain standard library classes like `URLSearchParams`

### ISSUE THREAD

**Issue by DetachHead** [typescript-eslint/typescript-eslint#7607]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/restrict-template-expressions/ and https://typescript-eslint.io/rules/restrict-plus-operands/

### Description

when `URLSearchParams` gets converted to a string, it turns into the string representation of the url parameters. this is what you want most of the time, so i think these rules should allow it (or at least have an option to allow it)

### Fail

```typescript
await fetch('/foo?' + {a: 1, b: 2})  // evaluates to "/foo?[object Object]"
```

### Pass

```typescript
await fetch('/foo?' + new URLSearchParams({a: 1, b: 2})) // evaluates to "/foo?a=1&b=2"
```

**Comment by bradzacher** [comment]
Please see the note on the docs here:
https://typescript-eslint.io/rules/restrict-template-expressions/

The rules purposely do not allow implicit string coercion for objects that define a `.toString` method because not all cases are going to be correct strings that you actually want to use.

A good example is arrays (see the above note).

If you want to use the type's `toString` implementation you can explicitly call it which clearly documents the intent of your code.
